File: src/codes/Users_Manual/advanced/custom_gain_0.py
from collections.abc import Callable

import numpy as np
from pyautd3 import Device, Drive, EmitIntensity, Geometry, Phase, Transducer, rad
from pyautd3.gain import Gain


class Focus(Gain):
    def __init__(self, point):
        self.point = np.array(point)

    def calc(self, _: Geometry) -> Callable[[Device], Callable[[Transducer], Drive]]:
        return Gain._transform(
            lambda dev: lambda tr: Drive(
                Phase(
                    float(np.linalg.norm(tr.position - self.point))
                    * dev.wavenumber
                    * rad
                ),
                EmitIntensity.maximum(),
            ),
        )

File: src/codes/Users_Manual/advanced/custom_gain_0.cs
~using AUTD3Sharp;
~using AUTD3Sharp.Utils;
~using AUTD3Sharp.Gain;
~using static AUTD3Sharp.Units;
using AUTD3Sharp.Derive;
~var g = new FocalPoint(Vector3.Zero);

[Gain]
public partial class FocalPoint
{
    private readonly Vector3 _point;

    public FocalPoint(Vector3 point)
    {
        _point = point;
    }

    private Func<Device, Func<Transducer, Drive>> Calc(Geometry geometry)
    {
        return Transform(dev => tr =>
        {
            var tp = tr.Position;
            var dist = (tp - _point).L2Norm;
            var phase = dist * dev.Wavenumber;
            return new Drive { Phase = new Phase(phase * rad), Intensity = EmitIntensity.Max };
        });
    }
}